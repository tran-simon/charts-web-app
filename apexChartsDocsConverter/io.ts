import { promises as fs } from 'fs';
// @ts-ignore
import { parse } from 'himalaya';
import { JSDOM } from 'jsdom';
import { isDocSection, isDocValue } from './convert';
import prettier from 'prettier'

export const generateCode = async (data: string, name: string) => {
  const res = `
import { ApexOptions } from 'apexcharts';
import { Options } from '../optionModel';

const ${name}: Options<ApexOptions['${name}']> = ${data};

export default ${name}; 
`;

  const options = (await prettier.resolveConfig('../.prettierrc.json')) || {
    'semi': true,
    'trailingComma': 'all',
    'singleQuote': true,
    'tabWidth': 2,
  };
  return prettier.format(res, {
    parser: 'babel',
    ...options
  })
};

export const stringify = (data: object): string => {
  const replacer = (key: string, v: any) => {
    if (isDocValue(v)) {
      if (v.options) {

        const selectOptions = v.options
          .map((option) =>
            `${option}: { value: '${option}', labelId: ${[...v.path, 'option', option].join('.')} }`)

        return `new SelectOptionField({ ${selectOptions} })`;
      }

      if (v.type === 'string') {
        return `new TextOptionField()`;
      }
      if (v.type === 'number') {
        return `new NumberOptionField()`;
      }
      if (v.type === 'boolean') {
        return `new BoolOptionField()`;
      }
      if (v.type === 'string | number' || v.type === 'number | string') {
        // TODO(https://github.com/tran-simon/pretty-charts/issues/5)
        return `new TextOptionField()`;
      }
      if (v.type === 'color'){
        // TODO(https://github.com/tran-simon/pretty-charts/issues/6)
        return `new TextOptionField()`;
      }
    } else if (isDocSection(v)) {
      return  v.children.reduce((acc, curr) => {
        acc[curr.name] = curr
        return acc;
      }, {} as any);
    }
    return v;
  };

  return JSON.stringify(data, replacer, 2).replaceAll('"', '');
};

/**
 * HTML file to js object
 * @param html document data
 */
export const parseHtml = (html: Document) => {
  const optionsDetails = html.getElementsByClassName('options-details').item(0);

  if (!optionsDetails) {
    throw 'Error while retrieving HTML element: element with class `options-details` not found.';
  }

  return parse(optionsDetails.innerHTML);
};
