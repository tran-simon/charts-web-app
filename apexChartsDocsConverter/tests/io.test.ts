import { generateCode, parseHtml, stringify } from '../io';
import * as assert from 'assert';
import { JSDOM } from 'jsdom';
import fs from 'fs/promises'


describe('io', ()=>{
  it('can parseHtml', async ()=>{
    const document = (await JSDOM.fromFile('tests/title.html')).window.document;
    expect(parseHtml(document)).toMatchSnapshot()
  })

  it('can stringify', async ()=>{
    const data = JSON.parse(await fs.readFile('tests/titleConvertedData.json', 'utf-8'))
    expect(stringify(data)).toMatchSnapshot()
  })

  it('can generate code', async ()=>{
    const data = JSON.parse(await fs.readFile('tests/titleConvertedData.json', 'utf-8'))
    expect(await generateCode(stringify(data), 'testGenerateCode')).toMatchSnapshot()
  })
})
