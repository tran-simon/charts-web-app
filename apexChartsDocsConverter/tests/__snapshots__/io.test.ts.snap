// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`io can generate code 1`] = `
"import { ApexOptions } from 'apexcharts';
import { Options } from '../optionModel';

const testGenerateCode: Options<ApexOptions['testGenerateCode']> = {
  text: new TextOptionField(),
  align: new SelectOptionField({
    left: { value: 'left', labelId: title.align.option.left },
    center: { value: 'center', labelId: title.align.option.center },
    right: { value: 'right', labelId: title.align.option.right },
  }),
  margin: new NumberOptionField(),
  offsetX: new NumberOptionField(),
  offsetY: new NumberOptionField(),
  floating: new BoolOptionField(),
  style: {
    fontSize: new TextOptionField(),
    fontWeight: new TextOptionField(),
    fontFamily: new TextOptionField(),
    color: new TextOptionField(),
  },
};

export default testGenerateCode;
"
`;

exports[`io can parseHtml 1`] = `
Array [
  Object {
    "attributes": Array [
      Object {
        "key": "class",
        "value": "chart-title",
      },
    ],
    "children": Array [
      Object {
        "attributes": Array [],
        "children": Array [
          Object {
            "content": "title",
            "type": "text",
          },
        ],
        "tagName": "h3",
        "type": "element",
      },
      Object {
        "content": " ",
        "type": "text",
      },
      Object {
        "attributes": Array [
          Object {
            "key": "id",
            "value": "text",
          },
          Object {
            "key": "class",
            "value": "anchor",
          },
        ],
        "children": Array [],
        "tagName": "a",
        "type": "element",
      },
      Object {
        "attributes": Array [],
        "children": Array [
          Object {
            "attributes": Array [],
            "children": Array [
              Object {
                "content": "text: String",
                "type": "text",
              },
            ],
            "tagName": "h3",
            "type": "element",
          },
          Object {
            "attributes": Array [],
            "children": Array [
              Object {
                "content": " Text to display as a title of chart",
                "type": "text",
              },
            ],
            "tagName": "p",
            "type": "element",
          },
        ],
        "tagName": "div",
        "type": "element",
      },
      Object {
        "content": " ",
        "type": "text",
      },
      Object {
        "attributes": Array [
          Object {
            "key": "id",
            "value": "align",
          },
          Object {
            "key": "class",
            "value": "anchor",
          },
        ],
        "children": Array [],
        "tagName": "a",
        "type": "element",
      },
      Object {
        "attributes": Array [],
        "children": Array [
          Object {
            "attributes": Array [],
            "children": Array [
              Object {
                "content": "align: String",
                "type": "text",
              },
            ],
            "tagName": "h3",
            "type": "element",
          },
          Object {
            "attributes": Array [],
            "children": Array [
              Object {
                "content": " Alignment of title relative to chart area. ",
                "type": "text",
              },
              Object {
                "attributes": Array [],
                "children": Array [],
                "tagName": "br",
                "type": "element",
              },
              Object {
                "content": " Possible Options ",
                "type": "text",
              },
              Object {
                "attributes": Array [],
                "children": Array [],
                "tagName": "br",
                "type": "element",
              },
            ],
            "tagName": "p",
            "type": "element",
          },
          Object {
            "attributes": Array [],
            "children": Array [
              Object {
                "attributes": Array [],
                "children": Array [
                  Object {
                    "content": "left",
                    "type": "text",
                  },
                ],
                "tagName": "li",
                "type": "element",
              },
              Object {
                "attributes": Array [],
                "children": Array [
                  Object {
                    "content": "center",
                    "type": "text",
                  },
                ],
                "tagName": "li",
                "type": "element",
              },
              Object {
                "attributes": Array [],
                "children": Array [
                  Object {
                    "content": "right",
                    "type": "text",
                  },
                ],
                "tagName": "li",
                "type": "element",
              },
            ],
            "tagName": "ul",
            "type": "element",
          },
          Object {
            "attributes": Array [],
            "children": Array [],
            "tagName": "p",
            "type": "element",
          },
        ],
        "tagName": "div",
        "type": "element",
      },
      Object {
        "content": " ",
        "type": "text",
      },
      Object {
        "attributes": Array [
          Object {
            "key": "id",
            "value": "margin",
          },
          Object {
            "key": "class",
            "value": "anchor",
          },
        ],
        "children": Array [],
        "tagName": "a",
        "type": "element",
      },
      Object {
        "attributes": Array [],
        "children": Array [
          Object {
            "attributes": Array [],
            "children": Array [
              Object {
                "content": "margin: Number",
                "type": "text",
              },
            ],
            "tagName": "h3",
            "type": "element",
          },
          Object {
            "attributes": Array [],
            "children": Array [
              Object {
                "content": " Vertical spacing around the title text",
                "type": "text",
              },
            ],
            "tagName": "p",
            "type": "element",
          },
        ],
        "tagName": "div",
        "type": "element",
      },
      Object {
        "content": " ",
        "type": "text",
      },
      Object {
        "attributes": Array [
          Object {
            "key": "id",
            "value": "offsetX",
          },
          Object {
            "key": "class",
            "value": "anchor",
          },
        ],
        "children": Array [],
        "tagName": "a",
        "type": "element",
      },
      Object {
        "attributes": Array [],
        "children": Array [
          Object {
            "attributes": Array [],
            "children": Array [
              Object {
                "content": "offsetX: Number",
                "type": "text",
              },
            ],
            "tagName": "h3",
            "type": "element",
          },
          Object {
            "attributes": Array [],
            "children": Array [
              Object {
                "content": " Sets the left offset for title text",
                "type": "text",
              },
            ],
            "tagName": "p",
            "type": "element",
          },
        ],
        "tagName": "div",
        "type": "element",
      },
      Object {
        "content": " ",
        "type": "text",
      },
      Object {
        "attributes": Array [
          Object {
            "key": "id",
            "value": "offsetY",
          },
          Object {
            "key": "class",
            "value": "anchor",
          },
        ],
        "children": Array [],
        "tagName": "a",
        "type": "element",
      },
      Object {
        "attributes": Array [],
        "children": Array [
          Object {
            "attributes": Array [],
            "children": Array [
              Object {
                "content": "offsetY: Number",
                "type": "text",
              },
            ],
            "tagName": "h3",
            "type": "element",
          },
          Object {
            "attributes": Array [],
            "children": Array [
              Object {
                "content": " Sets the top offset for title text",
                "type": "text",
              },
            ],
            "tagName": "p",
            "type": "element",
          },
        ],
        "tagName": "div",
        "type": "element",
      },
      Object {
        "content": " ",
        "type": "text",
      },
      Object {
        "attributes": Array [
          Object {
            "key": "id",
            "value": "floating",
          },
          Object {
            "key": "class",
            "value": "anchor",
          },
        ],
        "children": Array [],
        "tagName": "a",
        "type": "element",
      },
      Object {
        "attributes": Array [],
        "children": Array [
          Object {
            "attributes": Array [],
            "children": Array [
              Object {
                "content": "floating: Boolean",
                "type": "text",
              },
            ],
            "tagName": "h3",
            "type": "element",
          },
          Object {
            "attributes": Array [],
            "children": Array [
              Object {
                "content": " The floating option will take out the title text from the chart area and make it float on top of the chart.",
                "type": "text",
              },
            ],
            "tagName": "p",
            "type": "element",
          },
        ],
        "tagName": "div",
        "type": "element",
      },
      Object {
        "content": " ",
        "type": "text",
      },
      Object {
        "attributes": Array [
          Object {
            "key": "id",
            "value": "style",
          },
          Object {
            "key": "class",
            "value": "anchor",
          },
        ],
        "children": Array [],
        "tagName": "a",
        "type": "element",
      },
      Object {
        "attributes": Array [],
        "children": Array [
          Object {
            "attributes": Array [],
            "children": Array [
              Object {
                "content": "style",
                "type": "text",
              },
            ],
            "tagName": "h3",
            "type": "element",
          },
          Object {
            "content": " ",
            "type": "text",
          },
          Object {
            "attributes": Array [
              Object {
                "key": "id",
                "value": "fontSize",
              },
              Object {
                "key": "class",
                "value": "anchor",
              },
            ],
            "children": Array [],
            "tagName": "a",
            "type": "element",
          },
          Object {
            "attributes": Array [],
            "children": Array [
              Object {
                "attributes": Array [],
                "children": Array [
                  Object {
                    "content": "fontSize: String",
                    "type": "text",
                  },
                ],
                "tagName": "h3",
                "type": "element",
              },
              Object {
                "attributes": Array [],
                "children": Array [
                  Object {
                    "content": " Font Size of the title text",
                    "type": "text",
                  },
                ],
                "tagName": "p",
                "type": "element",
              },
            ],
            "tagName": "div",
            "type": "element",
          },
          Object {
            "content": " ",
            "type": "text",
          },
          Object {
            "attributes": Array [
              Object {
                "key": "id",
                "value": "fontWeight",
              },
              Object {
                "key": "class",
                "value": "anchor",
              },
            ],
            "children": Array [],
            "tagName": "a",
            "type": "element",
          },
          Object {
            "attributes": Array [],
            "children": Array [
              Object {
                "attributes": Array [],
                "children": Array [
                  Object {
                    "content": "fontWeight: String | Number",
                    "type": "text",
                  },
                ],
                "tagName": "h3",
                "type": "element",
              },
              Object {
                "attributes": Array [],
                "children": Array [
                  Object {
                    "content": " Font Weight of the title text",
                    "type": "text",
                  },
                ],
                "tagName": "p",
                "type": "element",
              },
            ],
            "tagName": "div",
            "type": "element",
          },
          Object {
            "content": " ",
            "type": "text",
          },
          Object {
            "attributes": Array [
              Object {
                "key": "id",
                "value": "fontFamily",
              },
              Object {
                "key": "class",
                "value": "anchor",
              },
            ],
            "children": Array [],
            "tagName": "a",
            "type": "element",
          },
          Object {
            "attributes": Array [],
            "children": Array [
              Object {
                "attributes": Array [],
                "children": Array [
                  Object {
                    "content": "fontFamily: String",
                    "type": "text",
                  },
                ],
                "tagName": "h3",
                "type": "element",
              },
              Object {
                "attributes": Array [],
                "children": Array [
                  Object {
                    "content": " Font Family of the title text",
                    "type": "text",
                  },
                ],
                "tagName": "p",
                "type": "element",
              },
            ],
            "tagName": "div",
            "type": "element",
          },
          Object {
            "content": " ",
            "type": "text",
          },
          Object {
            "attributes": Array [
              Object {
                "key": "id",
                "value": "color",
              },
              Object {
                "key": "class",
                "value": "anchor",
              },
            ],
            "children": Array [],
            "tagName": "a",
            "type": "element",
          },
          Object {
            "attributes": Array [],
            "children": Array [
              Object {
                "attributes": Array [],
                "children": Array [
                  Object {
                    "content": "color: Color",
                    "type": "text",
                  },
                ],
                "tagName": "h3",
                "type": "element",
              },
              Object {
                "attributes": Array [],
                "children": Array [
                  Object {
                    "content": " Fore color of the title text",
                    "type": "text",
                  },
                ],
                "tagName": "p",
                "type": "element",
              },
            ],
            "tagName": "div",
            "type": "element",
          },
        ],
        "tagName": "div",
        "type": "element",
      },
    ],
    "tagName": "div",
    "type": "element",
  },
]
`;

exports[`io can stringify 1`] = `
"{
  text: new TextOptionField(),
  align: new SelectOptionField({ left: { value: 'left', labelId: title.align.option.left },center: { value: 'center', labelId: title.align.option.center },right: { value: 'right', labelId: title.align.option.right } }),
  margin: new NumberOptionField(),
  offsetX: new NumberOptionField(),
  offsetY: new NumberOptionField(),
  floating: new BoolOptionField(),
  style: {
    fontSize: new TextOptionField(),
    fontWeight: new TextOptionField(),
    fontFamily: new TextOptionField(),
    color: new TextOptionField()
  }
}"
`;
